pkg/remote/resolver.go

```diff
@@ -17,22 +14,20 @@ limitations under the License.
 package remote
 
 import (
-	"github.com/google/go-containerregistry/pkg/authn"
-	"github.com/google/go-containerregistry/pkg/authn/k8schain"
-	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
 )
 
-// Resolver will retrieve Tekton resources like Tasks from remote repositories like an OCI image repositories.
-type Resolver interface {
-	GetTask(taskName string) (*v1beta1.TaskSpec, error)
+// ResolvedObject is returned by Resolver.List representing a Tekton resource stored in a remote location.
+type ResolvedObject struct {
+	Kind       string
+	APIVersion string
+	Name       string
 }
 
-// TODO: Right now, there is only one resolver type. When more are added, this will need to be updated.
-func NewResolver(imageReference string, serviceAccountName string) Resolver {
-	return OCIResolver{
-		imageReference: imageReference,
-		keychainProvider: func() (authn.Keychain, error) {
-			return k8schain.NewInCluster(k8schain.Options{ServiceAccountName: serviceAccountName})
-		},
-	}
+// Resolver defines a generic API to retrieve Tekton resources from remote locations. It allows 2 principle operations:
+//   - List:     retrieve a flat set of Tekton objects in this remote location
+//   - Get:      retrieves a specific object with the given Kind and name.
+type Resolver interface {
```

Comment:
This interface would be better defined where we use it (and we *might* not even need to be exported in that case).
One reason for this is the dependencies it brings if we want to use it — aka any package the `pkg/remote` package depends on.

This means at least:
```
	"github.com/google/go-containerregistry/pkg/authn"
	imgname "github.com/google/go-containerregistry/pkg/name"
	v1 "github.com/google/go-containerregistry/pkg/v1"
	"github.com/google/go-containerregistry/pkg/v1/remote"
	"github.com/tektoncd/pipeline/pkg/client/clientset/versioned/scheme"
	"k8s.io/apimachinery/pkg/runtime"
```

- Using this interface *shouldn't* require any dependency on `go-containerregistry` (using an impl. might).
- I wonder about the `runtime.Object` return case too :thinking:

Reply:
Re: `runtime.Object` it was the most generic, universal object I could think of to return either a Pipeline, Task, or ClusterTask out of the same API.

On the larger question of where to put the interface. I was thinking it would make sense to co-locate it with the definitions since that is where users would expect it to be. You are correct that it would require bringing in the container dependencies and if we add a git implementation, git dependencies as well. I guess we could put the `resolver` interface under the `remote` package to centralize it and then create sub-packages with each implementation so that you only import the ones you care about? Does that make sense?

Reply:
> On the larger question of where to put the interface. I was thinking it would make sense to co-locate it with the definitions since that is where users would expect it to be. You are correct that it would require bringing in the container dependencies and if we add a git implementation, git dependencies as well. I guess we could put the `resolver` interface under the `remote` package to centralize it and then create sub-packages with each implementation so that you only import the ones you care about? Does that make sense?

I would work — it's not the best from my point of view (aka I am not a huge fan of exported interface), but I think it's a good start :wink:

Reply:
Done